pkg/controllers/nodeclass/status/instanceprofile.go

```diff
@@ -0,0 +1,50 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
+)
+
+type InstanceProfile struct {
+	instanceProfileProvider instanceprofile.Provider
+}
+
+func (ip *InstanceProfile) Reconcile(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (reconcile.Result, error) {
+	if controllerutil.ContainsFinalizer(nodeClass, v1beta1.TerminationFinalizer) {
```

Comment:
Can this just be added at the top-level for now and patched at the beginning of the reconcile. This code still has a race since you can still go and create the instance profile, but fail to update the Kubernetes object

Reply:
Moved it to the top level of the reconcile loop. Seems like this race is currently present, no? Since we can fail to update the k8s object even in the current NodeClass controller